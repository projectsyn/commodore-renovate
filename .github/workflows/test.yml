name: Test
on:
  pull_request:
    branches:
      - master

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Lint
        run: |
          yarn lint
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Setup Python
        uses: actions/setup-python@v4
        with:
          python-version: '3.9'
      - name: Cache pip
        uses: actions/cache@v3
        with:
          # This path is specific to Ubuntu
          path: ~/.cache/pip
          # Look to see if there is a cache hit for the corresponding requirements file
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
      - name: Install dependencies
        run: pip install -r requirements.txt
      - name: Test
        run: |
          yarn test

  docker-build:
    runs-on: ubuntu-latest
    env:
      IMAGE: ${{ github.repository }}
    steps:
      - uses: actions/checkout@v4
      - name: Get changed files
        id: changed-files
        uses: tj-actions/changed-files@v39
      - name: Build
        if: contains(steps.changed-files.outputs.modified_files, 'Dockerfile')
        uses: docker/build-push-action@v4
        with:
          push: false
          tags: 'ghcr.io/${{ env.IMAGE }}:test'
