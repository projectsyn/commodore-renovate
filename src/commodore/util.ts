import clone from 'just-clone';
import yaml from 'js-yaml';
import Git from 'simple-git';

import { parse } from 'path';
import { mkdir, readFile, writeFile } from 'fs/promises';
import { URL } from 'url';

import { logger } from 'renovate/dist/logger';
import { getGlobalConfig } from 'renovate/dist/config/global';
import { simpleGitConfig } from 'renovate/dist/util/git/config';

import type { Facts, RepoConfig } from './types';

export function cacheDir(): string {
  let cacheDir: string | undefined = getGlobalConfig().cacheDir;
  if (cacheDir === undefined) {
    cacheDir = '/tmp/renovate';
  }
  return cacheDir;
}

/* Prune object, removing any null or undefined keys */
export function pruneObject(obj: Facts) {
  return Object.entries(obj)
    .filter(([_, v]) => v !== undefined && v !== null)
    .reduce((main, [key, value]) => ({ ...main, [key]: value }), {});
}

/* Render object as YAML and write to file */
export async function writeYamlFile(
  filePath: string,
  obj: object
): Promise<void> {
  const objYaml: string = yaml.dump(obj);
  return writeFile(filePath, objYaml);
}

/* Extract distribution/cloud/region from filename */
export function parseFileName(
  fileName: string,
  distributionRegex: RegExp,
  cloudRegionRegex: RegExp,
  ignoreValues: Array<string>
): Facts {
  const distA = fileName.match(distributionRegex);
  const cloudA = fileName.match(cloudRegionRegex);

  const dist = distA ? distA.groups : null;
  const cloud = cloudA ? cloudA.groups : null;

  let facts: Facts = {
    distribution: null,
    cloud: null,
    region: null,
  };
  if (dist !== undefined && dist !== null) {
    facts.distribution = dist.distribution;
    facts.cloud = dist.cloud ? dist.cloud : null;
  }
  if (cloud !== undefined && cloud !== null) {
    facts.cloud = cloud.cloud;
    facts.region = cloud.region ? cloud.region : null;
  }

  if (
    facts.distribution !== null &&
    ignoreValues.includes(facts.distribution)
  ) {
    facts.distribution = null;
  }
  if (facts.cloud !== null && ignoreValues.includes(facts.cloud)) {
    facts.cloud = null;
  }
  if (facts.region !== null && ignoreValues.includes(facts.region)) {
    facts.region = null;
  }

  return facts;
}

function isObject(item: any) {
  return item && typeof item === 'object' && !Array.isArray(item);
}

// Adapted from https://stackoverflow.com/a/37164538
export function mergeConfig(base: any, config: any): any {
  let output = clone(base);
  if (isObject(base) && isObject(config)) {
    Object.keys(config).forEach((key) => {
      if (isObject(config[key])) {
        if (!(key in base)) Object.assign(output, { [key]: config[key] });
        else output[key] = mergeConfig(base[key], config[key]);
      } else if (Array.isArray(config[key])) {
        if (!(key in base)) {
          Object.assign(output, { [key]: Array.from(config[key]) });
        } else if (Array.isArray(output[key])) {
          output[key].push(...config[key]);
        } else {
        }
      } else {
        Object.assign(output, { [key]: config[key] });
      }
    });
  }
  return output;
}

export var globalRepos: Map<string, RepoConfig> = new Map();

export function globalRepoDir(globalRepoURL: string): string {
  let repoUrl = new URL(globalRepoURL);
  let repoPathInfo = parse(repoUrl.pathname);
  let repoPath = `${repoUrl.host}${repoPathInfo.dir}/${repoPathInfo.name}`;
  return cacheDir() + `/global-repos/${repoPath}`;
}

export async function cloneGlobalRepo(config: any): Promise<RepoConfig> {
  if (globalRepos.has(config.globalRepoURL)) {
    return globalRepos.get(config.globalRepoURL) as RepoConfig;
  }

  logger.info(
    {
      globalRepoURL: config.globalRepoURL,
      tenantId: config.tenantId,
    },
    'Global repo for this tenant is not initialized yet, cloning it'
  );
  const dir: string = globalRepoDir(config.globalRepoURL);
  await mkdir(dir, { recursive: true });
  // TODO(sg): support Git-https?
  const git: any = Git(dir, simpleGitConfig());
  await git.clone(config.globalRepoURL, '.');
  const globalExtraConfig = await parseGlobalRepoConfig(dir);
  const rc: RepoConfig = {
    dir: dir,
    extraConfig: globalExtraConfig,
  } as RepoConfig;
  globalRepos.set(config.globalRepoURL, rc);
  return rc;
}

export async function parseGlobalRepoConfig(globalDir: string): Promise<any> {
  const globalDirConfigFile: string = globalDir + '/renovate.json';
  let globalDirConfig: any = {};
  try {
    const globalConfigStr: string = (
      await readFile(globalDirConfigFile)
    ).toString();
    globalDirConfig = JSON.parse(globalConfigStr) as any;
  } catch (e: any) {
    logger.info(
      {
        globalDir: globalDir,
        exception: e,
      },
      'Exception while reading global repo config'
    );
  }
  if (
    globalDirConfig.commodore != null &&
    globalDirConfig.commodore.extraConfig != null
  ) {
    try {
      const globalExtraConfigStr = (
        await readFile(globalDir + '/' + globalDirConfig.commodore.extraConfig)
      ).toString();
      return JSON.parse(globalExtraConfigStr) as any;
    } catch (e: any) {
      logger.info(
        {
          globalDir: globalDir,
          extraConfig: globalDirConfig.commodore.extraConfig,
          exception: e,
        },
        'Exception while reading global repo extraConfig'
      );
    }
  }
  return {};
}
