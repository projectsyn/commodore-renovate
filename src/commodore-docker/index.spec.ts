import { loadFixture } from '../test/util';
import { extractPackageFile, parseImageDependency } from './index';
import { expect, describe, it } from '@jest/globals';

const params1 = loadFixture('1/params.yml');
const params2 = loadFixture('2/params.yml');
const params3 = loadFixture('3/params.yml');

describe('manager/commodore/index', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      expect(extractPackageFile('nothing here', 'no.yml')).toBeNull();
    });
    it('extracts standard images', () => {
      const res = extractPackageFile(params1, '1/params.yml');
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(4);
      }
    });
    it('returns no images for files without images', () => {
      expect(extractPackageFile(params2, '2/params.yml')).toBeNull();
    });
    it('returns images for non standard but common formats', () => {
      const res = extractPackageFile(params3, '3/params.yml');
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(4);
      }
    });
  });
});


describe('manager/commodore/index/parseImageDependency', () => {
  const table = [
    {
      image: {
        registry: 'quay.io',
        repository: 'bitnami/kubectl',
        tag: '1.21.2',
      },
      depName: 'quay.io/bitnami/kubectl',
      currentValue: '1.21.2',
    },
    {
      image: {
        repository: 'quay.io/bitnami/kubectl',
        tag: '1.21.2',
      },
      depName: 'quay.io/bitnami/kubectl',
      currentValue: '1.21.2',
    },
    {
      image: {
        registry: 'quay.io',
        image: 'bitnami/kubectl',
        tag: '1.21.2',
      },
      depName: 'quay.io/bitnami/kubectl',
      currentValue: '1.21.2',
    },
    {
      image: {
        registry: 'quay.io',
        image: 'bitnami/kubectl',
        version: '1.21.2',
      },
      depName: 'quay.io/bitnami/kubectl',
      currentValue: '1.21.2',
    },
    {
      image: {
        image: 'bitnami/kubectl:1.21.2',
      },
      noImage: true,
    },
    {
      image: {
        registry: 'quay.io',
        tag: '1.21.2',
      },
      noImage: true,
    },
    {
      image: 'blub',
      noImage: true,
    },
    {
      noImage: true,
    },
  ]
  describe.each(table)('parseImageDependency', (tc) => {
    if (tc?.noImage) {
      expect(parseImageDependency(tc.image)).toBeNull()
    } else {
      expect(parseImageDependency(tc.image)?.depName).toBe(tc.depName)
      expect(parseImageDependency(tc.image)?.currentValue).toBe(tc.currentValue)
    }
  })
})
