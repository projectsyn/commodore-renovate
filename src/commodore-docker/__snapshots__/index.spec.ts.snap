// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`manager/commodore/index > extractPackageFile() > extracts standard images 1`] = `
[
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "1.21.2",
    "datasource": "docker",
    "depName": "quay.io/bitnami/kubectl",
    "packageName": "quay.io/bitnami/kubectl",
    "replaceString": "1.21.2",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "1.1.2",
    "datasource": "docker",
    "depName": "quay.io/projectsyn/foo",
    "packageName": "quay.io/projectsyn/foo",
    "replaceString": "1.1.2",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.8.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/bar",
    "packageName": "docker.io/projectsyn/bar",
    "replaceString": "v0.8.0",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/buzz",
    "packageName": "docker.io/projectsyn/buzz",
    "replaceString": "v0.6.0",
    "versioning": "docker",
  },
]
`;

exports[`manager/commodore/index > extractPackageFile() > returns images for non standard but common formats 1`] = `
[
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/steward",
    "packageName": "docker.io/projectsyn/steward",
    "replaceString": "v0.6.0@sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "currentValue": "v2.0.4",
    "datasource": "docker",
    "depName": "quay.io/argoproj/argocd",
    "packageName": "quay.io/argoproj/argocd",
    "replaceString": "v2.0.4@sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "15.0.2",
    "datasource": "docker",
    "depName": "quay.io/keycloak/keycloak",
    "packageName": "quay.io/keycloak/keycloak",
    "replaceString": "15.0.2",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/buzz",
    "packageName": "docker.io/projectsyn/buzz",
    "replaceString": "v0.6.0",
    "versioning": "docker",
  },
]
`;

exports[`manager/commodore/index > extractPackageFile() > returns images that are deeply nested in each other 1`] = `
[
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/steward",
    "packageName": "docker.io/projectsyn/steward",
    "replaceString": "v0.6.0@sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "currentValue": "v2.0.4",
    "datasource": "docker",
    "depName": "quay.io/argoproj/argocd",
    "packageName": "quay.io/argoproj/argocd",
    "replaceString": "v2.0.4@sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "15.0.2",
    "datasource": "docker",
    "depName": "quay.io/keycloak/keycloak",
    "packageName": "quay.io/keycloak/keycloak",
    "replaceString": "15.0.2",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/buzz",
    "packageName": "docker.io/projectsyn/buzz",
    "replaceString": "v0.6.0",
    "versioning": "docker",
  },
]
`;

exports[`manager/commodore/index > extractPackageFile() > returns images that are not declared in parameters.<component>.images 1`] = `
[
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/steward",
    "packageName": "docker.io/projectsyn/steward",
    "replaceString": "v0.6.0@sha256:36468cfc3cdd3877e80b08a68426530a1bcd5a17fe03bfc301e63027ded30272",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "currentValue": "v2.0.4",
    "datasource": "docker",
    "depName": "quay.io/argoproj/argocd",
    "packageName": "quay.io/argoproj/argocd",
    "replaceString": "v2.0.4@sha256:976dfbfadb817ba59f4f641597a13df7b967cd5a1059c966fa843869c9463348",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "15.0.2",
    "datasource": "docker",
    "depName": "quay.io/keycloak/keycloak",
    "packageName": "quay.io/keycloak/keycloak",
    "replaceString": "15.0.2",
    "versioning": "docker",
  },
  {
    "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.6.0",
    "datasource": "docker",
    "depName": "docker.io/projectsyn/buzz",
    "packageName": "docker.io/projectsyn/buzz",
    "replaceString": "v0.6.0",
    "versioning": "docker",
  },
]
`;
